lesson_2/rps_bonus.py

```diff
@@ -0,0 +1,109 @@
+'''' Update Rock, Paper, Scissors Program where you play against the computer:
+add Lizard & Spock to choices:
+    Rules:
+    scissors beats: paper & lizard
+    paper beats: rock & spock
+    rock beats: lizard & scissors
+    lizard beats: Spock & paper
+    spock beats: scissors & rock
+'''
+
+import random
+
+VALID_CHOICES = { 
+    'rock' : {'rock', 'r'},
+    'paper' : {'paper', 'p'},
+    'scissors' : {'scissors', 'sc'},
+    'lizard' : {'lizard', 'l'},
+    'spock' : {'spock', 'sp'}
+    }
+
+def choice_invalid(choice):
+    for value in VALID_CHOICES.values():
+        for element in value:
+            if choice != element:
+                continue
+            else:
+                if choice == element:
+                    return False # choice is valid
+
+def update_choice(choice):            
+    for key in VALID_CHOICES.keys():
+        if choice in VALID_CHOICES[key]:
+            update_choice = key
+            return update_choice
+
+def prompt(message):
+    print(f"==> {message}")
+
+def display_error(message):
+    print(f"    !!!! ERROR: Invalid input. {message} !!!!")
+
+def display_welcome():
+    print('\n----------------------------------------')
+    print('ROCK * PAPER * SCISSORS * LIZARD * SPOCK')
+    print('----------------GO!---------------------')
+
+# TODO fix ERROR results not matching rules
+#   CHECK IF THESE REALLY ARE THE ONLY OPTIONS
+#  scissors beats: paper & lizard
+#     paper beats: rock & spock
+#     rock beats: lizard & scissors
+#     lizard beats: Spock & paper
+#     spock beats: scissors & rock
+def display_winner(user, computer):
```

Comment:
Here's a cleaner and more maintainable way to write this function:

```py
def display_winner(user, computer):
    winning_combinations = {
        'rock': ['scissors', 'lizard'],
        'paper': ['rock', 'spock'],
        'scissors': ['paper', 'lizard'],
        'lizard': ['paper', 'spock'],
        'spock': ['scissors', 'rock']
    }

    if computer in winning_combinations.get(user, []):
        display_results('YOU WIN! :D')
    elif user in winning_combinations.get(computer, []):
        display_results('COMPUTER WINS!')
    else:
        display_results("IT'S A TIE!")
```

In this case, if you need to add a new winning combination you just update `winning_combinations` instead of updating it in two places via the `if/else` blocks.

Reply:
love to see it